File: app-play/app/assets/javascripts/profile/ConnectController.js
(function () {
    var app = angular.module('shocktrade');

    /**
     * Connect Controller
     * @author lawrence.daniels@gmail.com
     */
    app.controller('ConnectController', ['$scope', '$log', 'toaster', 'ComposeMessageDialog', 'ConnectService', 'MySession',
        function ($scope, $log, toaster, ComposeMessageDialog, ConnectService, MySession) {
            $scope.searchName = "";
            $scope.myUpdates = [];
            $scope.myUpdate = null;
            $scope.contact = {};

            /**
             * Selects a specific contact/friend
             */
            $scope.chooseContact = function (friend) {
                //$log.info("contact = " + JSON.stringify(friend, null, '\t'));
                $scope.contact = friend;
                $scope.getUserInfo(friend.id);
            };

            /**
             * Selects the contact in the player's friends list
             */
            $scope.chooseFirstContact = function () {
                var friends = MySession.fbFriends;
                if (friends && friends.length > 0) {
                    $scope.chooseContact(friends[0]);
                }
            };

            /**
             * Retrieve a limited set of user profile information for a specific contact/friend
             */
            $scope.getUserInfo = function (fbUserId) {
                $scope.startLoading();
                ConnectService.getUserInfo(fbUserId).then(
                    function (profile) {
                        $scope.stopLoading();
                        $scope.contact.profile = profile;
                    },
                    function (err) {
                        $scope.stopLoading();
                        console.log("Error: Connect -" + data + " (" + status + ")");
                        toaster.pop('error', 'Failed to retrieve the user profile for contact ' + $scope.contact.name, null);
                    }
                );
            };

            $scope.identifyFacebookFriends = function (fbFriends) {
                $scope.startLoading();
                ConnectService.identifyFacebookFriends(fbFriends).then(
                    function (response) {
                        // TODO add logic here
                        $scope.stopLoading();
                    },
                    function (err) {
                        $scope.stopLoading();
                    });
            };

            /**
             * Returns the contacts matching the given search term
             */
            $scope.getContactList = function (searchTerm) {
                var fbFriends = MySession.fbFriends;
                if (!searchTerm && (searchTerm === "")) return fbFriends.slice(0, 40);
                else {
                    var term = searchTerm.toLowerCase();
                    return fbFriends.filter(function (contact) {
                        return contact.name.toLowerCase().indexOf(term) != -1;
                    });
                }
            };

            /**
             * Loads updates from the database
             */
            $scope.loadMyUpdates = function (userName) {
                if (userName) {
                    $scope.startLoading();

                    ConnectService.getUserUpdates(userName, 50)
                        .success(function (data) {
                            $scope.stopLoading();
                            $scope.myUpdates = data;
                            $scope.myUpdate = null;

                            for (var n = 0; n < data.length; n++) {
                                data[n].selected = false;
                            }
                            $scope.loading = false;
                        })
                        .error(function (err) {
                            $scope.stopLoading();
                            console.log("Error: Connect -" + data + " (" + status + ")");
                            toaster.pop('error', "Failed to load Connect", null);
                            $scope.loading = false;
                        });
                }
            };

            /**
             * Selects a specific message
             */
            $scope.selectUpdate = function (entry) {
                $scope.myUpdate = entry;
            };

            /**
             * Composes a new message via pop-up dialog
             */
            $scope.composeMessage = function () {
                ComposeMessageDialog.popup({
                    "success": function() {
                        $scope.loadMyUpdates(MySession.getUserName());
                    }
                });
            };

            /**
             * Deletes selected messages
             */
            $scope.deleteMessages = function (userName) {
                $scope.startLoading();

                // gather the records to delete
                var myUpdates = $scope.myUpdates;
                var messageIDs = [];
                for (var n = 0; n < myUpdates.length; n++) {
                    if (myUpdates[n].selected) {
                        messageIDs.push(myUpdates[n].OID());
                    }
                }

                // delete the records
                $log.info("messageIDs = " + angular.toJson(messageIDs));
                if (messageIDs.length > 0) {
                    ConnectService.deleteMessages(messageIDs).then(
                        function (response) {
                            $scope.stopLoading();
                            $scope.loadMyUpdates(userName);
                        },
                        function (err) {
                            $scope.stopLoading();
                            toaster.pop('error', 'Failed to delete message', null);
                        }
                    );
                }
                else {
                    $scope.stopLoading();
                    toaster.pop('error', 'No message(s) selected', null);
                }
            };

            /**
             * Selects all currently visible messages
             */
            $scope.selectAll = function (checked) {
                var data = $scope.myUpdates;
                for (var n = 0; n < data.length; n++) {
                    data[n].selected = checked;
                }
            };

            // watch for changes to the player's profile
            $scope.$watch(MySession.getUserProfile(), function () {
                $scope.loadMyUpdates(MySession.getUserName());
                //$scope.chooseFirstContact();
            });

        }]);

})();

File: app-js/src/main/scala/com/shocktrade/javascript/profile/ConnectController.scala
package com.shocktrade.javascript.profile

import biz.enef.angulate.core.Timeout
import biz.enef.angulate.{Scope, ScopeController, named}
import com.ldaniels528.angularjs.Toaster
import com.shocktrade.javascript.MySession
import com.shocktrade.javascript.ScalaJsHelper._

import scala.scalajs.js
import scala.scalajs.js.Dynamic.{global => g, literal => JS}
import scala.scalajs.js.JSON
import scala.util.{Failure, Success}

/**
 * Connect Controller
 * @author lawrence.daniels@gmail.com
 */
class ConnectController($scope: js.Dynamic, toaster: Toaster,
                        @named("ComposeMessageDialog") composeMessageDialog: js.Dynamic,
                        @named("ConnectService") connectService: ConnectService,
                        @named("MySession") mySession: MySession)
  extends ScopeController {

  private val scope = $scope.asInstanceOf[Scope]
  private var myUpdates = emptyArray[js.Dynamic]
  private var myUpdate: js.Dynamic = null
  private var contact: js.Dynamic = JS()

  /**
   * Selects a specific contact/friend
   */
  $scope.chooseContact = (friend: js.Dynamic) => {
    //g.console.log(s"contact = ${JSON.stringify(friend, null, "\t")}")
    contact = friend
    $scope.getUserInfo(friend.id)
  }

  /**
   * Selects the contact in the player"s friends list
   */
  $scope.chooseFirstContact = () => mySession.fbFriends.headOption foreach ($scope chooseContact _)

  $scope.getContact = () => contact

  $scope.getMyUpdates = () => myUpdates

  $scope.getSelectedUpdate = () => myUpdate

  /**
   * Selects a specific message
   */
  $scope.selectUpdate = (entry: js.Dynamic) => myUpdate = entry

  /**
   * Retrieve a limited set of user profile information for a specific contact/friend
   */
  $scope.getUserInfo = (fbUserId: String) => {
    $scope.startLoading()
    connectService.getUserInfo(fbUserId) onComplete {
      case Success(profile) =>
        $scope.stopLoading()
        contact.profile = profile
      case Failure(e) =>
        $scope.stopLoading()
        g.console.log(s"Failed to retrieve profile for contact ${contact.name}")
        toaster.error(s"Failed to retrieve the user profile for contact ${contact.name}")
    }
  }

  $scope.identifyFacebookFriends = (fbFriends: js.Array[js.Dynamic]) => {
    $scope.startLoading()
    connectService.identifyFacebookFriends(fbFriends) onComplete {
      case Success(_) => $scope.stopLoading()
      case Failure(e) => $scope.stopLoading()
    }
  }

  /**
   * Returns the contacts matching the given search term
   */
  $scope.getContactList = (searchTerm: js.UndefOr[String]) => {
    // TODO reinstate search
    /*
    g.console.log(s"searchTerm = $searchTerm (${Option(searchTerm).map(_.getClass.getName).orNull})")
    searchTerm foreach { mySearchTerm =>
      val fbFriends = mySession.fbFriends
      if (!Option(searchTerm).exists(_.nonEmpty)) fbFriends.take(40)
      else {
        val term = mySearchTerm.trim.toLowerCase
        fbFriends.filter(friend => isDefined(friend.name) && friend.name.as[String].toLowerCase.contains(term))
      }
    }*/
    mySession.fbFriends
  }

  /**
   * Loads updates from the database
   */
  $scope.loadMyUpdates = (userName: String) => loadMyUpdates(userName)

  private def loadMyUpdates(userName: String) = {
    if (userName.nonBlank) {
      $scope.startLoading()

      connectService.getUserUpdates(userName, 50) onComplete {
        case Success(data) =>
          $scope.stopLoading()
          myUpdates = data
          myUpdate = null
          data.foreach(_.selected = false)
          $scope.loading = false

        case Failure(e) =>
          $scope.stopLoading()
          g.console.log("Failed to load Connect")
          e.printStackTrace()
          toaster.pop("error ", "Failed to load Connect", null)
          $scope.loading = false
      }
    }
  }

  /**
   * Composes a new message via pop-up dialog
   */
  $scope.composeMessage = () => composeMessageDialog.popup(JS(success = () => loadMyUpdates(mySession.getUserName())))

  /**
   * Deletes selected messages
   */
  $scope.deleteMessages = (userName: String) => {
    $scope.startLoading()

    // gather the records to delete
    val messageIDs = myUpdates.filter(update => isDefined(update.selected) && update.selected.as[Boolean]).map(_.OID)

    // delete the records
    g.console.log(s"messageIDs = ${JSON.stringify(messageIDs)}")
    if (messageIDs.nonEmpty) {
      connectService.deleteMessages(messageIDs) onComplete {
        case Success(response) =>
          $scope.stopLoading()
          loadMyUpdates(userName)

        case Failure(e) =>
          $scope.stopLoading()
          toaster.error("Failed to delete message ")
      }
    }
    else {
      $scope.stopLoading()
      toaster.error("No message(s) selected")
    }
  }

  /**
   * Selects all currently visible messages
   */
  $scope.selectAll = (checked: Boolean) => myUpdates.foreach(_.selected = checked)

  // watch for changes to the player"s profile
  scope.$watch(mySession.getUserID, () => {
    loadMyUpdates(mySession.getUserName())
    //$scope.chooseFirstContact()
  })

}
